run as administration when holding ctrl and shift

themike10452 opened the pull request:
add run as administration when holding ctrl and shift to Program and Explorer plugins. 

change Shell plugin's admin key from ctrl to ctrl and shift

taooceros commented on Flow.Launcher/MainWindow.xaml:
```diff
@@ -46,6 +46,7 @@
         <KeyBinding Key="Home" Modifiers="Alt" Command="{Binding SelectFirstResultCommand}"></KeyBinding>
         <KeyBinding Key="O" Modifiers="Ctrl" Command="{Binding LoadContextMenuCommand}"></KeyBinding>
         <KeyBinding Key="H" Modifiers="Ctrl" Command="{Binding LoadHistoryCommand}"></KeyBinding>
+        <KeyBinding Key="Enter" Modifiers="Ctrl+Shift" Command="{Binding OpenResultCommand}"></KeyBinding>
```
Is this needed?

themike10452 replied:
yes since CTRL+Enter and Shift+Enter are already bound each on its own, if you press both CTRL and Shift the command will not be triggered unless we add an explicit binding for both being pressed at the same time.

taooceros commented on Plugins/Flow.Launcher.Plugin.Explorer/Search/ResultManager.cs:
```diff
@@ -127,7 +129,26 @@ internal static Result CreateFileResult(string filePath, Query query, int score
-                        if (c.SpecialKeyState.CtrlPressed)
+                        if (File.Exists(filePath) && c.SpecialKeyState.CtrlPressed && c.SpecialKeyState.ShiftPressed)
+                        {
+                            Task.Run(() =>
```
Is adding `Task.Run` help responsibility?

themike10452 replied:
yes it does, I tested this using a batch file that wait for 10 seconds before exiting. Without Task.Run the UI freezes for 10 seconds until the batch file is done executing then it hides itself. With Task.Run the UI hides itself immediately and is ready to be used again while the batch file is still running.

taooceros reviewed (commented):
Thanks for contribution!

jjw24 commented:
Thanks @themike10452

What about changing Ctrl + Shift to be just Ctrl? So it's the same as Shell plugin invoking administrator.

themike10452 commented:
@jjw24 i wanted to mimic the behavior of windows start menu where ctrl+shift+enter will run the selected program as administrator. Microsoft's PowerToys Run uses the same key combination as well.

jjw24 commented:
> @jjw24 i wanted to mimic the behavior of windows start menu where ctrl+shift+enter will run the selected program as administrator. Microsoft's PowerToys Run uses the same key combination as well.

Yeah ok I didn't know that's the windows default. 

@taooceros worth realign the default admin key combo for shell plugin?

taooceros commented:
> > @jjw24 i wanted to mimic the behavior of windows start menu where ctrl+shift+enter will run the selected program as administrator. Microsoft's PowerToys Run uses the same key combination as well.
> 
> Yeah ok I didn't know that's the windows default.
> 
> @taooceros worth realign the default admin key combo for shell plugin?

@themike10452 Do you want to handle that in this pr together?

themike10452 commented:
> > > @jjw24 i wanted to mimic the behavior of windows start menu where ctrl+shift+enter will run the selected program as administrator. Microsoft's PowerToys Run uses the same key combination as well.
> > 
> > 
> > Yeah ok I didn't know that's the windows default.
> > @taooceros worth realign the default admin key combo for shell plugin?
> 
> @themike10452 Do you want to handle that in this pr together?

sure I'll add that to the PR. I wanted to look at launching UWP apps as admin as well.

themike10452 commented:
@jjw24 I added 2 commits to this PR, we can now launch UWP apps and shell commands as admin when holding ctrl + shift.
I also added a run as admin option to the context menu of UWP app results.
Programs must be reindexed before they can run as admin.

taooceros commented:
Amazing, I want to launcher windows terminal as administrative for a long time.

themike10452 commented:
> Amazing, I want to launcher windows terminal as administrative for a long time.

me too!

taooceros commented on Plugins/Flow.Launcher.Plugin.Program/Programs/UWP.cs:
```diff
@@ -320,7 +321,27 @@ public Result Result(string query, IPublicAPI api)
+                        {
+                            if (!CanRunElevated)
+                            {
+                                return false;
```
Maybe notify user that this app cannot be run as admin (and probably launch it as normal user).

themike10452 replied:
done

taooceros commented on Plugins/Flow.Launcher.Plugin.Program/Programs/UWP.cs:
```diff
@@ -377,6 +413,20 @@ private async void Launch(IPublicAPI api)
 
+            private async void LaunchElevated()
+            {
+                string command = "shell:AppsFolder\\" + UniqueIdentifier;
+                command = Environment.ExpandEnvironmentVariables(command.Trim());
+
+                var info = new ProcessStartInfo(command)
+                {
+                    UseShellExecute = true,
+                    Verb = "runas",
+                };
+
+                Main.StartProcess(Process.Start, info);
```
if this works, do we still need to activation helper for uwp app?

themike10452 replied:
not really, however, Microsoft's launcher uses the activation helper for normal user launch and this method for elevated launch.

taooceros commented on Plugins/Flow.Launcher.Plugin.Program/Programs/UWP.cs:
```diff
@@ -403,6 +453,31 @@ public Application(AppxPackageHelper.IAppxManifestApplication manifestApp, UWP p
+                {
+                    return true;
+                }
+                else
```
maybe remove the else will be clearer

themike10452 replied:
done

taooceros reviewed (dismissed):
Will wait for a while to merge for others to take a look

jjw24 reviewed (commented):
LGTM, thank you @themike10452. Just added version bumps since we also added admin launch for files in Explorer plugin.

jjw24 commented:
@taooceros let's hold for 1.9.0. 

This pr is available in the pre-release https://github.com/Flow-Launcher/Flow.Launcher/discussions/640

themike10452 commented on Plugins/Flow.Launcher.Plugin.Program/Languages/tr.xaml:
```diff
@@ -35,5 +35,6 @@
     <system:String x:Key="flowlauncher_plugin_program_plugin_description">Programları Flow Launcher'tan arayın</system:String>
 
     <system:String x:Key="flowlauncher_plugin_program_invalid_path">Geçersiz Konum</system:String>
+    <system:String x:Key="flowlauncher_plugin_program_run_as_administrator_not_supported_message">This app is not intended to be run as administrator</system:String>
```
that's fine I'll remove them. I did notice other strings that are not translated so I did the same for the new string I added.

themike10452 commented on Plugins/Flow.Launcher.Plugin.Program/Programs/UWP.cs:
```diff
@@ -320,7 +321,29 @@ public Result Result(string query, IPublicAPI api)
+                            e.SpecialKeyState.CtrlPressed &&
+                            e.SpecialKeyState.ShiftPressed &&
+                            !e.SpecialKeyState.AltPressed &&
+                            !e.SpecialKeyState.WinPressed
```
because ctrl+alt+shift is not the same as ctrl+shift... in case we wanted to add a different action for ctrl+alt+shift we can with this check in place.
